### PR TITLE
Fix for dual certificate issue MAISTRA-1142

### PR DESCRIPTION
**Pull request by twghu**
Description: Fix for RSA/ECDSA dual cert configuration and tls_context issue.
Risk Level:
Testing: Tested ssl_integration_test
Docs Changes: none.
Release Notes: 

[reproducer.zip](https://github.com/Maistra/envoy/files/4388122/reproducer.zip)

Fixes #Issue  MAISTRA-1142

### CODE REVIEW

File: source/extensions/transport_sockets/tls/context_impl.cc [hunk 1]
```diff
@@ -46,7 +46,9 @@ ContextImpl::ContextImpl(Stats::Scope& scope, const Envoy::Ssl::ContextConfig& c
       ssl_curves_(stat_name_set_->add("ssl.curves")),
       ssl_sigalgs_(stat_name_set_->add("ssl.sigalgs")) {
   const auto tls_certificates = config.tlsCertificates();
-  tls_contexts_.resize(std::max(static_cast<size_t>(1), tls_certificates.size()));
+// Use a single context for certificates instead of one context per certificate as
+// openssl handles the multi cert case automatically in the handshake
+  tls_contexts_.resize(1);
```
Comment (dmitri-d) [inline source/extensions/transport_sockets/tls/context_impl.cc]: if the context is always of size 1, why not drop the vector altogether?

Comment (twghu) [reply source/extensions/transport_sockets/tls/context_impl.cc]: Yes I agree.

I had a look at the context-impl.cc and there are a few lines I want to be sure of before taking that route.

Comment (twghu) [reply source/extensions/transport_sockets/tls/context_impl.cc]: There are a number of places in the context-impl.cc where tls_contexts_ are referenced in for loops, it should be low risk to dispose of the vector.

Comment (twghu) [reply source/extensions/transport_sockets/tls/context_impl.cc]: Agreed with Dmitri we will drop the vector in the server context.

**Comment by twghu** [comment]
The commit 4d541bd removes the reliance on the multiple TlsContext (ie a vector) ContextImpl class.
OpenSSL manages the handshake with the certificates and does not required the callbacks used for BoringSSL.

File: source/extensions/transport_sockets/tls/context_impl.cc [hunk 2]
```diff
@@ -46,11 +46,9 @@ ContextImpl::ContextImpl(Stats::Scope& scope, const Envoy::Ssl::ContextConfig& c
       ssl_curves_(stat_name_set_->add("ssl.curves")),
       ssl_sigalgs_(stat_name_set_->add("ssl.sigalgs")) {
   const auto tls_certificates = config.tlsCertificates();
-// Use a single context for certificates instead of one context per certificate as
-// openssl handles the multi cert case automatically in the handshake
-  tls_contexts_.resize(1);
 
-  for (auto& ctx : tls_contexts_) {
+  auto& ctx = tls_context_;
+  {
```
Comment (dmitri-d) [inline source/extensions/transport_sockets/tls/context_impl.cc]: Here and below: could you either remove the braces left after ```for``` loops, or add TODOs to deal with them later?

Comment (twghu) [reply source/extensions/transport_sockets/tls/context_impl.cc]: Originally left the {} in place to minimise risk of failure, on further analysis removed and re-tested.

File: source/extensions/transport_sockets/tls/context_impl.cc [hunk 3]
```diff
@@ -101,8 +99,8 @@ ContextImpl::ContextImpl(Stats::Scope& scope, const Envoy::Ssl::ContextConfig& c
                                        config.certificateValidationContext()->caCertPath()));
     }
 
-    for (auto& ctx : tls_contexts_) {
-      X509_STORE* store = SSL_CTX_get_cert_store(ctx.ssl_ctx_.get());
+    {
```
Comment (dmitri-d) [inline source/extensions/transport_sockets/tls/context_impl.cc]: Could you remove extra braces here?

File: source/extensions/transport_sockets/tls/context_impl.cc [hunk 4]
```diff
@@ -158,8 +156,8 @@ ContextImpl::ContextImpl(Stats::Scope& scope, const Envoy::Ssl::ContextConfig& c
                       config.certificateValidationContext()->certificateRevocationListPath()));
     }
 
-    for (auto& ctx : tls_contexts_) {
-      X509_STORE* store = SSL_CTX_get_cert_store(ctx.ssl_ctx_.get());
+    {
```
Comment (dmitri-d) [inline source/extensions/transport_sockets/tls/context_impl.cc]: Could you remove extra braces here?

**Comment by dmitri-d** [comment]
@twghu: Could you re-enable MultiCertPreferEcdsa test in ```test/extensions/transport_sockets/tls/ssl_socket_test.cc``` please?

**Pull request merged**